# Data Preprocessing 

In [2]:
#importing libraries for data preprocessing
import pandas as pd
import numpy as np

In [27]:
df = pd.read_csv("/content/adult.csv") #reading dataset

In [28]:
df.head() #shows top 5 rows of dataset

age workclass  fnlwgt  ... hours.per.week  native.country income
0   90         ?   77053  ...             40   United-States  <=50K
1   82   Private  132870  ...             18   United-States  <=50K
2   66         ?  186061  ...             40   United-States  <=50K
3   54   Private  140359  ...             40   United-States  <=50K
4   41   Private  264663  ...             40   United-States  <=50K

[5 rows x 15 columns]

In [29]:
df.shape 

(32561, 15)

shape shows that dataset contains 24918 columns and 15 rows

In [30]:
df.describe() 

age        fnlwgt  ...  capital.loss  hours.per.week
count  32561.000000  3.256100e+04  ...  32561.000000    32561.000000
mean      38.581647  1.897784e+05  ...     87.303830       40.437456
std       13.640433  1.055500e+05  ...    402.960219       12.347429
min       17.000000  1.228500e+04  ...      0.000000        1.000000
25%       28.000000  1.178270e+05  ...      0.000000       40.000000
50%       37.000000  1.783560e+05  ...      0.000000       40.000000
75%       48.000000  2.370510e+05  ...      0.000000       45.000000
max       90.000000  1.484705e+06  ...   4356.000000       99.000000

[8 rows x 6 columns]

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


# **Checking, If Any Column Contains Null Values**


In [22]:
for col in df:
  print(col.upper(), "Column contains: \n")
  print(df[col].value_counts(),'\n\n\n')
 

AGE Column contains: 

36    683
37    677
31    676
23    674
33    672
     ... 
82      9
84      8
83      6
88      3
85      2
Name: age, Length: 71, dtype: int64 



WORKCLASS Column contains: 

Private             17257
Self-emp-not-inc     1991
Local-gov            1641
?                    1376
State-gov             978
Self-emp-inc          897
Federal-gov           765
Without-pay             8
Never-worked            5
Name: workclass, dtype: int64 



FNLWGT Column contains: 

203488    12
126675    11
241998    10
102308    10
186934    10
          ..
277444     1
63296      1
96337      1
632593     1
249541     1
Name: fnlwgt, Length: 17764, dtype: int64 



EDUCATION Column contains: 

HS-grad         7999
Some-college    5546
Bachelors       4213
Masters         1358
Assoc-voc       1065
11th             878
Assoc-acdm       791
10th             698
7th-8th          479
Prof-school      474
9th              377
Doctorate        329
12th             310
5th-6th      

In [32]:
#Replacing '?' with 'NaN' values
df.replace('?', np.NaN, inplace = True)
df.head()

age workclass  fnlwgt  ... hours.per.week  native.country income
0   90       NaN   77053  ...             40   United-States  <=50K
1   82   Private  132870  ...             18   United-States  <=50K
2   66       NaN  186061  ...             40   United-States  <=50K
3   54   Private  140359  ...             40   United-States  <=50K
4   41   Private  264663  ...             40   United-States  <=50K

[5 rows x 15 columns]

In [33]:
#Applying Forward Fill to Replace 'NaN'
df.fillna(method = 'ffill', inplace = True)
df.head()

age workclass  fnlwgt  ... hours.per.week  native.country income
0   90       NaN   77053  ...             40   United-States  <=50K
1   82   Private  132870  ...             18   United-States  <=50K
2   66   Private  186061  ...             40   United-States  <=50K
3   54   Private  140359  ...             40   United-States  <=50K
4   41   Private  264663  ...             40   United-States  <=50K

[5 rows x 15 columns]

Here we found that
Workclass column has 1376 values from unknown category.
Occupation column has 1381 values from unknown category.
Native Country column has 449 values from unknown country.

In [34]:
#Dropping Unuseful data
df.drop(['education.num', 'fnlwgt','capital.gain', 'capital.loss'], axis = 1, inplace=True)
df.head()

age workclass     education  ... hours.per.week native.country income
0   90       NaN       HS-grad  ...             40  United-States  <=50K
1   82   Private       HS-grad  ...             18  United-States  <=50K
2   66   Private  Some-college  ...             40  United-States  <=50K
3   54   Private       7th-8th  ...             40  United-States  <=50K
4   41   Private  Some-college  ...             40  United-States  <=50K

[5 rows x 11 columns]

We Assumed that educatio.num, fnlwgt, capital.gain, capital.loss columns are not affecting our data more so we drop these.

# **Lab**el **Encoding**
Here we will apply label encoding for changing categorical values into numerical because machine understands only numbers.

In [26]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['workclass'] = le.fit_transform(df['workclass'])
df['marital.status'] = le.fit_transform(df['marital.status'])
df['occupation'] = le.fit_transform(df['occupation'])
df['relationship'] = le.fit_transform(df['relationship'])
df['sex'] = le.fit_transform(df['sex'])
df['race'] = le.fit_transform(df['race'])
df['native.country'] = le.fit_transform(df['native.country'])
df['income'] = le.fit_transform(df['income'])
df.head()

age  workclass     education  ...  hours.per.week  native.country  income
0   90          8       HS-grad  ...            40.0              38       0
1   82          3       HS-grad  ...            18.0              38       0
2   66          3  Some-college  ...            40.0              38       0
3   54          3       7th-8th  ...            40.0              38       0
4   41          3  Some-college  ...            40.0              38       0

[5 rows x 11 columns]

In [35]:
df = df.apply(LabelEncoder().fit_transform)
df.head()

age  workclass  education  ...  hours.per.week  native.country  income
0   72          8         11  ...              39              38       0
1   65          3         11  ...              17              38       0
2   49          3         15  ...              39              38       0
3   37          3          5  ...              39              38       0
4   24          3         15  ...              39              38       0

[5 rows x 11 columns]

In [36]:
x = df.drop(['income'], axis = 1)   #Dropping column ['income'] 
y = df['income']    #Assigning column ['income'] to y or using as a class

# Data Split

In [49]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.2) #70% data as training_data and 30% as testing_data

In [38]:
x_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22792 entries, 21136 to 4986
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   age             22792 non-null  int64
 1   workclass       22792 non-null  int64
 2   education       22792 non-null  int64
 3   marital.status  22792 non-null  int64
 4   occupation      22792 non-null  int64
 5   relationship    22792 non-null  int64
 6   race            22792 non-null  int64
 7   sex             22792 non-null  int64
 8   hours.per.week  22792 non-null  int64
 9   native.country  22792 non-null  int64
dtypes: int64(10)
memory usage: 1.9 MB


In [50]:
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB
gnb = GaussianNB() #Creating object for Gaussian Naive Bayes
gnb.fit(x_train, y_train) #fitting data into Gaussian Naive Bayes

bnb = BernoulliNB() #Creating object for Bernoulli Naive Bayes
bnb.fit(x_train, y_train) #fitting data into Bernoulli Naive Bayes

mnb = MultinomialNB() #Creating object for Multinomial Naive Bayes
mnb.fit(x_train, y_train) #fitting data into Multinomial Naive Bayes

MultinomialNB()

In [45]:
#Performance Computing modules
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

In [53]:
#Gausssian Naive Bayes
y_pred = gnb.predict(x_test)
print("Classification Report: \n",classification_report(y_test, y_pred))
print("Confusion Matrix is:\n ",confusion_matrix(y_test, y_pred))
print("Accuracy is: ",accuracy_score(y_test, y_pred))

Classification Report: 
               precision    recall  f1-score   support

           0       0.89      0.78      0.83      4978
           1       0.50      0.70      0.58      1535

    accuracy                           0.76      6513
   macro avg       0.70      0.74      0.71      6513
weighted avg       0.80      0.76      0.78      6513

Confusion Matrix is:
  [[3899 1079]
 [ 462 1073]]
Accuracy is:  0.7633962843543682


In [54]:
#Bernoulli Naive Bayes
y_pred = bnb.predict(x_test)
print("Classification Report:\n",classification_report(y_test, y_pred))
print("Confusion Matrix is:\n ",confusion_matrix(y_test, y_pred))
print("Accuracy is: ",accuracy_score(y_test, y_pred))

Classification Report:
               precision    recall  f1-score   support

           0       0.90      0.73      0.81      4978
           1       0.46      0.74      0.57      1535

    accuracy                           0.73      6513
   macro avg       0.68      0.74      0.69      6513
weighted avg       0.80      0.73      0.75      6513

Confusion Matrix is:
  [[3652 1326]
 [ 402 1133]]
Accuracy is:  0.7346844771994473


In [56]:
#Multinomial Naive Bayes
y_pred = mnb.predict(x_test)
print("Classification Report:\n ",classification_report(y_test, y_pred))
print("Confusion Matrix is:\n ",confusion_matrix(y_test, y_pred))
print("Accuracy is: ",accuracy_score(y_test, y_pred))

Classification Report:
                precision    recall  f1-score   support

           0       0.85      0.73      0.78      4978
           1       0.40      0.59      0.48      1535

    accuracy                           0.69      6513
   macro avg       0.63      0.66      0.63      6513
weighted avg       0.75      0.69      0.71      6513

Confusion Matrix is:
  [[3616 1362]
 [ 625  910]]
Accuracy is:  0.6949178565945033
